core/src/banking_stage/worker.rs

```diff
@@ -0,0 +1,631 @@
+use {
+    super::{
+        consumer::{Consumer, ExecuteAndCommitTransactionsOutput, ProcessTransactionBatchOutput},
+        forwarder::Forwarder,
+        scheduler_messages::{ConsumeWork, FinishedConsumeWork, FinishedForwardWork, ForwardWork},
+        ForwardOption,
+    },
+    crossbeam_channel::{select, Receiver, RecvError, SendError, Sender},
+    solana_poh::leader_bank_notifier::LeaderBankNotifier,
+    solana_runtime::bank::Bank,
+    std::{sync::Arc, time::Duration},
+    thiserror::Error,
+};
+
+#[derive(Debug, Error)]
+pub enum WorkerError {
+    #[error("Failed to receive work from scheduler: {0}")]
+    Recv(#[from] RecvError),
+    #[error("Failed to send finalized consume work to scheduler: {0}")]
+    ConsumedSend(SendError<FinishedConsumeWork>),
+    #[error("Failed to send finalized forward work to scheduler: {0}")]
+    ForwardedSend(SendError<FinishedForwardWork>),
+}
+
+impl From<SendError<FinishedConsumeWork>> for WorkerError {
+    fn from(err: SendError<FinishedConsumeWork>) -> Self {
+        Self::ConsumedSend(err)
+    }
+}
+
+impl From<SendError<FinishedForwardWork>> for WorkerError {
+    fn from(err: SendError<FinishedForwardWork>) -> Self {
+        Self::ForwardedSend(err)
+    }
+}
+
+pub(crate) struct Worker {
+    consume_receiver: Receiver<ConsumeWork>,
+    consumer: Consumer,
+    consumed_sender: Sender<FinishedConsumeWork>,
+
+    forward_receiver: Receiver<ForwardWork>,
+    forward_option: ForwardOption,
+    forwarder: Forwarder,
+    forwarded_sender: Sender<FinishedForwardWork>,
+
+    leader_bank_notifier: Arc<LeaderBankNotifier>,
+}
+
+#[allow(dead_code)]
+impl Worker {
+    pub fn new(
+        consume_receiver: Receiver<ConsumeWork>,
+        consumer: Consumer,
+        consumed_sender: Sender<FinishedConsumeWork>,
+        forward_receiver: Receiver<ForwardWork>,
+        forward_option: ForwardOption,
+        forwarder: Forwarder,
+        forwarded_sender: Sender<FinishedForwardWork>,
+        leader_bank_notifier: Arc<LeaderBankNotifier>,
+    ) -> Self {
+        Self {
+            consume_receiver,
+            consumer,
+            consumed_sender,
+            forward_receiver,
+            forward_option,
+            forwarder,
+            forwarded_sender,
+            leader_bank_notifier,
+        }
+    }
+
+    pub fn run(self) -> Result<(), WorkerError> {
+        loop {
+            select! {
+                recv(self.consume_receiver) -> work => {
+                    self.consume_loop(work?)?;
+                },
+                recv(self.forward_receiver) -> work => {
+                    self.forward_loop(work?)?;
+                },
+            }
```

Comment:
thanks to all the hard refactoring by you, this succinct main loop body is embodiment of it. :)

that progress made me to question this: why a worker is responsible for both consuming and forwarding? why not create separate worker threads in the arrangement of 4 consumer worker threads, 2 forward worker threads (assuming forwarding must be network bound)?

i know both consume and forward don't happen at the same time, but i don't see a particular merit of doing the both inside a single worker thread either, creating a potential blocking source for each other. 

is it still intentional (if so, how about commenting in source?) or just leftover from the current banking stage impl?

Reply:
That's a very fair criticism. I'd thought about separating them and chose not to because there's no expectation of forwarding and consuming to happen simultaneously.

However, you're correct. If we had a lot of queued forwarding work it could delay the start of consume! We've already got separate channels, I'll make the split!

Reply:
Opposite is also true, lots of queued consume work could delay forwarding...though I think this case is not as problematic

Reply:
Sorry for this commit that splits forward/consume workers: abc870fd729964ae7d6f73d28f8585e4f8f7d279

Essentially erases your previous diff views.